### PR TITLE
Enable Wcns5z reconstruction to use adaptive fallback

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes

* Delete Wcns5zMc reconstruction and related codes from develop branch.
* Extend the Wcns5z reconstructor in FD numerical algorithms with an additional template parameter `FallbackReconstructor` to enable adaptive reconstruction.
* Fix the Wcns5z class in ValenciaDivClean/FiniteDifference to match the updated Wcns5z reconstructor.

First commit is from:
- [ ] #4057 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/FiniteDifference/Wcns5z.cpp [hunk 1]
```diff
@@ -12,43 +12,91 @@
...
+               void (*)(gsl::not_null<DataVector*>, const DataVector&,
+                        const DataVector&, const Index<Dim>&, const Index<Dim>&,
+                        const Direction<Dim>&, const double&, const size_t&),
+               void (*)(gsl::not_null<DataVector*>, const DataVector&,
+                        const DataVector&, const Index<Dim>&, const Index<Dim>&,
+                        const Direction<Dim>&, const double&, const size_t&)>;
+
+// returns functions pointer
+template <size_t NonlinearWeightExponent, typename Reconstructor, size_t Dim>
+pointer_return_type<Dim> function_pointer_v() {
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/FiniteDifference/Wcns5z.cpp]: why is there a `_v` suffix? This isn't a constexpr variable, which is where `_v` is normally used

File: src/NumericalAlgorithms/FiniteDifference/Wcns5z.hpp [hunk 2]
```diff
@@ -154,44 +197,54 @@ struct Wcns5zReconstructor {
  * The nonlinear weights exponent \f$q (=1 \text{ or } 2)\f$ and the small
  * factor \f$\varepsilon\f$ can be chosen via an input file.
  *
+ * If the template parameter `FallbackReconstructor` is set to one of
+ * the FD reconstructor structs of which names are listed in
+ * `fd::reconstruction::FallbackReconstructorType`, adaptive reconstruction
+ * is performed as follows. For each finite difference stencils first check how
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/FiniteDifference/Wcns5z.hpp]: `stencils first` -> `stencil, first`

File: src/NumericalAlgorithms/FiniteDifference/Wcns5z.hpp [hunk 3]
```diff
@@ -154,44 +197,54 @@ struct Wcns5zReconstructor {
...
+ * many extrema are in the stencil. If the number of local extrema is less than
+ * or equal to a non-negative integer `max_number_of_extrema` which is given as
+ * an input parameter, perform the WCNS-5Z reconstruction; otherwise switch to
+ * the given `FallbackReconstructor` for performing reconstruction. If
+ * `FallbackReconstructor` is set to `void`, the adaptive method is disabled and
+ * always uses Wcns5z reconstruction.
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/FiniteDifference/Wcns5z.hpp]: `and ...` -> `and WCNS-5Z is always used.`

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/Wcns5z.hpp [hunk 4]
```diff
@@ -75,8 +79,25 @@ class Wcns5zPrim : public Reconstructor {
         "The parameter added to the oscillation indicators to avoid division "
         "by zero"};
   };
+  struct FallbackReconstructor {
+    using type = FallbackReconstructorType;
+    static constexpr Options::String help = {
+        "A reconstruction scheme to fallback adaptively. Finite difference "
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/Wcns5z.hpp]: `fallback` -> `fallback to`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
you can squash immediately

File: src/NumericalAlgorithms/FiniteDifference/Wcns5z.cpp (hunk 1 above)
Comment (yoonso0-0) [reply src/NumericalAlgorithms/FiniteDifference/Wcns5z.cpp]: Good to know. I'll remove this suffix

**Comment by yoonso0-0** [comment]
Squashed and rebased

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**